engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -84,68 +57,127 @@ class EngineJobExecutionActor(jobKey: BackendJobDescriptorKey,
-      backendJobExecutionActor ! message
-      context.parent ! JobRunning(jobDescriptor, backendJobExecutionActor)
-      goto(RunningJob)
-    case Event(response: BackendJobPreparationFailed, _) =>
+    case Event(BackendJobPreparationSucceeded(jobDescriptor, bjeaProps), NoData) =>
+      callCachingMode.activity match {
+        case Some(activity) if activity.readFromCache =>
+          initialiseJobHashing(jobDescriptor, activity)
+          goto(CheckingCallCache) using EJEAJobDescriptorData(Option(jobDescriptor), Option(bjeaProps))
+        case Some(activity) =>
+          initialiseJobHashing(jobDescriptor, activity)
+          runJob(jobDescriptor, bjeaProps)
+        case None => runJob(jobDescriptor, bjeaProps)
+      }
+    case Event(response: BackendJobPreparationFailed, NoData) =>
       context.parent forward response
       context stop self
       stay()
   }
 
+  // When CheckingCallCache, the FSM always has EJEAJobDescriptorData
+  when(CheckingCallCache) {
+    case Event(CacheMiss, EJEAJobDescriptorData(Some(jobDescriptor), Some(bjeaProps))) =>
+      log.info(s"Cache miss for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}")
+      runJob(jobDescriptor, bjeaProps)
+    case Event(CacheHit(cacheResultId), EJEAJobDescriptorData(Some(jobDescriptor), _)) =>
+      log.info(s"Cache hit for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}! Copying cache result $cacheResultId")
+      lookupCachedResult(jobDescriptor, cacheResultId)
+  }
+
+  // When RunningJob, the FSM always has EJEAPartialCompletionData (which might be None, None)
   when(RunningJob) {
-    case Event(response: BackendJobExecutionResponse, _) =>
-      saveJobCompletionResult(response)
-      context.parent forward response
-      goto(AwaitingJobStoreAck)
+    case Event(response: SucceededResponse, EJEAPartialCompletionData(None, Some(hashes))) if callCachingMode.writeToCache =>
+      saveCacheResults(EJEACompletionDataWithHashes(response, hashes))
+    case Event(response: SucceededResponse, data @ EJEAPartialCompletionData(None, None)) if callCachingMode.writeToCache =>
+      stay using data.copy(jobResult = Option(response))
+    case Event(hashes: CallCacheHashes, data @ EJEAPartialCompletionData(Some(response), None)) =>
+      saveCacheResults(EJEACompletionDataWithHashes(response, hashes))
+    case Event(hashes: CallCacheHashes, data @ EJEAPartialCompletionData(None, None)) =>
+      stay using data.copy(hashes = Option(hashes))
+    case Event(response: BackendJobExecutionResponse, data: EJEAPartialCompletionData) =>
+      saveJobCompletionToJobStore(response)
   }
 
-  when(AwaitingJobStoreAck) {
-    case Event(JobStoreWriteSuccess(_), _) =>
+  // When WritingToCallCache, the FSM always has EJEACompletionDataWithHashes
+  when(WritingToCallCache) {
+    // TODO: case Event(CacheWriter.SuccessAck, ...) =>
```

Comment:
Just dbl checking that these TODOs (here and throughout) are TODO for the call cache work, and not some fuzzy time way in the future